### PR TITLE
A sample for gRPC Server in custom model

### PR DESCRIPTION
**Pull request by Iamlovingit**
# A sample for gRPC Server in custom model

This sample can be used to try out gRPC, HTTP/2, and custom port configuration
in a custom inference  service.

## Build and Deploy the sample code

1. Use Docker to build a container image for this service and push to Docker Hub.

  Replace `{username}` with your Docker Hub username then run the commands:

  ```shell
  # Build the container on your local machine.
  docker build -t {username}/helloworld-grpc .

  # Push the container to docker registry.
  docker push {username}/helloworld-grpc
  ```

3. Update the `grpc-service.yaml` file in the project to reference the published image from step 1.

   Replace `{username}` in `grpc-service.yaml` with your Docker Hub user name:

   

  ```yaml
 apiVersion: serving.kubeflow.org/v1alpha2
kind: InferenceService
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: grpc-service
spec:
  default:
    predictor:
      custom:
        container:
          image: docker.io/{username}/helloworld-grpc:latest
          ports:
          - name: h2c
            containerPort: 8080
  ```

4. Use `kubectl` to deploy the service.

  ```shell
  kubectl apply -f grpc-service.yaml
  ```

## Testing the service

**Get the inference service address:**

```
kubectl get ksvc grpc-service-predictor-default
```

**Expect output**

```
NAME                             URL                                                                  READY   REASON
grpc-service-predictor-default   http://grpc-service-predictor-default.default.1.2.3.4.xip.io   True    
```

**How to setup the magic DNS**

 [Install Istio for Knative](https://knative.dev/docs/install/installing-istio/), at the `Configuring DNS` section, it describe how to config the DNS. 

You can also take a look at this [sample](https://github.com/kubeflow/kfserving/tree/master/docs/samples#deploy-kfserving-inferenceservice-with-a-custom-predictor) for more `custom model` detail.

Testing the gRPC service requires using a gRPC client built from the same
protobuf definition used by the server.

The Dockerfile builds the client binary. To run the client you will use the
same container image deployed for the server with an override to the
entrypoint command to use the client binary instead of the server binary.

Replace `{username}` with your Docker Hub user name and run the command:

```shell
docker run --rm {username}/helloworld-grpc \
  /client \
  -server_addr="grpc-service-predictor-default.default.1.2.3.4.xip.io:80" \
  -insecure
```

The arguments after the container tag `{username}/helloworld-grpc are used
instead of the entrypoint command defined in the Dockerfile `CMD` statement.

**Expect output**

```
2020/06/18 09:35:13 SyaHello got Hello world
2020/06/18 09:35:13 SendSomething got Hello KFServing
```

### CODE REVIEW

**Comment by Iamlovingit** [comment]
match issue #877

**Comment by yuzisun** [comment]
this is great example! Thanks @Iamlovingit 

/ok-to-test

File: docs/samples/custom/grpc-server/client/client.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+var (
+	serverAddr         = flag.String("server_addr", "127.0.0.1:8080", "The server address in the format of host:port")
+	serverHostOverride = flag.String("server_host_override", "", "")
+	insecure           = flag.Bool("insecure", false, "Set to true to skip SSL validation")
+	skipVerify         = flag.Bool("skip_verify", false, "Set to true to skip server hostname verification in SSL validation")
+)
+
+func main() {
```
Comment (yuzisun) [inline docs/samples/custom/grpc-server/client/client.go:26]: Do you have a simple model serving gRPC example? this looks like the knative gRPC example :)

Comment (Iamlovingit) [reply docs/samples/custom/grpc-server/client/client.go:26]: yes, It is the knative grpc example. Is it neccesarry to change a example?

File: docs/samples/custom/grpc-server/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# Copyright 2018 The Knative Authors
```
Comment (jinchihe) [inline docs/samples/custom/grpc-server/Dockerfile:1]: update to "2020 The Kubeflow Authors" ?

File: docs/samples/custom/grpc-server/client/client.go (hunk 1 above)
Comment (Iamlovingit) [reply docs/samples/custom/grpc-server/client/client.go:24]: I have updated the sample code to a simple helloworld server.

File: docs/samples/custom/grpc-server/Dockerfile (hunk 2 above)
Comment (Iamlovingit) [reply docs/samples/custom/grpc-server/Dockerfile]: I have updated the sample code to a simple helloworld server.

File: docs/samples/custom/grpc-server/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+```
+kubectl get ksvc grpc-service-predictor-default
+```
+
+**Expect output**
+
+```
+NAME                             URL                                                                  READY   REASON
+grpc-service-predictor-default   http://grpc-service-predictor-default.default.1.2.3.4.xip.io   True
```
Comment (yuzisun) [inline docs/samples/custom/grpc-server/README.md:61]: Can you add the pointer how this magic dns is setup?

Comment (Iamlovingit) [reply docs/samples/custom/grpc-server/README.md:61]: yes, I have added.

File: docs/samples/custom/grpc-server/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+```
+
+**Expect output**
+
+```
+NAME                             URL                                                                  READY   REASON
+grpc-service-predictor-default   http://grpc-service-predictor-default.default.1.2.3.4.xip.io   True    
+```
+**How to setup the magic dns**
+[Install Istio for Knative](https://knative.dev/docs/install/installing-istio/)
```
Comment (yuzisun) [inline docs/samples/custom/grpc-server/README.md:64]: Can you mention that it is the `Configure DNS` section and also link this sample here https://github.com/kubeflow/kfserving/tree/master/docs/samples#deploy-kfserving-inferenceservice-with-a-custom-predictor ? thanks!

Comment (Iamlovingit) [reply docs/samples/custom/grpc-server/README.md]: OK

**Comment by yuzisun** [comment]
Great writeup! thanks @Iamlovingit !
/lgtm
/approve

**Pull request merged**